### DISCUSSION TITLE
Unit tests and CI/CD (Actions)

### DISCUSSION THREAD

**Discussion by supcik** [Bodmer/TFT_eSPI#2200] [General]
Is there any CI/CD to make sure that TFT_eSPI works as expected? I did not find "Actions" configured for this repository, but there is maybe something else.

**Comment by Bodmer** [comment]
No, I find most problems relate to new features and need real hardware to test the library, displayed image fidelity and corner cases.

**Reply by supcik** [reply]
Thank you for your reply. For some projects, I am using another library for Arduino ([arduino-lmic](https://github.com/mcci-catena/arduino-lmic)) and in this project, a GitHub Action goes through all examples and ensures that they compile. I understand that many tests requires real hardware, but I think that having the examples to compile properly is a valuable first step.

Let me know if you support this idea and if so, I would be happy to contribute.

**Reply by Bodmer** [reply]
Thank you for your support offer.

I think that, due to the use of setup configuration files to support multiple processors and display variants the test script would be complicated. There are processor specific features (like DMA and RP2040 Programable IO), display variants, GPIO pin allocations and use of features built into some board packages (like SPIFFS and/or LittleFS for some processors) that will complicate matters.

Maybe these can be handled easily. The library has evolved organically over many years and ease of handling CI/CD has not been factored in.

What is your assessment of these considerations?

**Reply by supcik** [reply]
I understand, and I don't know the library good enough to make a legitimate judgement on the feasibility of good testing. For me, testing a little is better than no test at all, so I would start with simple basic tests and see how it evolves. Perhaps the community is kind enough to contribute once that the foundation is in place.

I am an avid supporter of beautiful code, and I am using CI to make sure that the style guidelines are enforced (using linters and formatters such as cpplint and clang-format). Going even further, I am using [pre-commit](https://pre-commit.com/) to check these rules even before pushing to the repository. It is possible to start small as well and expand as needed. I really like how [esp-idf](https://github.com/espressif/esp-idf) are checking the contributions.

But I don't want to interfere too much in your code and your way of managing this great project. You know, I teach computer sciences at a university and I am certainly influenced by my work ;-) so forgive me if I am too intrusive.

Anyhow, I really like your work and your way of maintaining this project. It's not often that you have someone who responds so quickly to messages, so thank you for your work.

**Reply by Bodmer** [reply]
I will think on this further!
